### PR TITLE
Add cabal-paths patch for ghc 9.2.x

### PR DESCRIPTION
**Pull request by sloane-shark**
###### Description of changes

Continuation of #184041 

I tweaked the cabal-paths.patch file included for GHC builds (see https://github.com/NixOS/nixpkgs/issues/140774) to work on the updated Cabal PathsModule code that is used in GHC 9.2.3+.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/ghc/cabal-paths-9.2.x.patch [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+     }
+   where
++    dirs = [ (flat_libdir, "LibDir")
++           , (flat_dynlibdir, "DynLibDir")
++           , (flat_datadir, "DataDir")
++           , (flat_libexecdir, "LibexecDir")
++           , (flat_sysconfdir, "SysconfDir") ];
++    shouldEmitPath p
++      | (splitPath flat_prefix) `isPrefixOf` (splitPath flat_bindir) = True
++      | (splitPath flat_prefix) `isPrefixOf` (splitPath p) = False
```
Comment (matthewess) [inline pkgs/development/compilers/ghc/cabal-paths-9.2.x.patch:46]: @sternenseemann Continuing the review from the last PR, what do you think is the proper fix here? I can't work out why splitPath was included originally (I copied this from the existing cabal-paths.patch file), but I imagine the correct code would be one of the following:
```suggestion
+      | flat_prefix `isPrefixOf` flat_bindir = True
+      | flat_prefix `isPrefixOf` p = False
```
```suggestion
+      | [flat_prefix] `isPrefixOf` (splitPath flat_bindir) = True
+      | [flat_prefix] `isPrefixOf` (splitPath p) = False
```

**Review by sloane-shark — COMMENTED** [review]
Changes had been requested on the last version of this PR, #184041

**Comment by matthewess** [comment]
Confirmed that `haskell.compilers.{ghc924,ghc925}` both build.

`jacinda` still fails, by way of `regex-rure`.

I've tried `nix run .#haskell.packages.{ghc924,ghc925}.{ormolu,ghcid.bin}`, and while both failed, it was for two different reasons, neither related to the error that occurs without the patch.

File: pkgs/development/compilers/ghc/cabal-paths-9.2.x.patch (hunk 1 above)
Comment (sternenseemann) [reply pkgs/development/compilers/ghc/cabal-paths-9.2.x.patch:46]: Yes, I think so.

**Comment by matthewess** [comment]
Result of `nixpkgs-review pr 216857` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>55 packages marked as broken and skipped:</summary>
  <ul>
    <li>cedille</li>
    <li>diagrams-builder</li>
    <li>emacsPackages.cedille</li>
    <li>emacsPackages.shm</li>
    <li>ghostwriter</li>
    <li>gssdp-tools</li>
    <li>gssdp_1_6</li>
    <li>hledger-iadd</li>
    <li>hyper-haskell-server-with-packages</li>
    <li>icepeak</li>
    <li>ldgallery</li>
    <li>madlang</li>
    <li>napari</li>
    <li>nix-delegate</li>
    <li>nix-linter</li>
    <li>notcurses</li>
    <li>petrinizer</li>
    <li>python310Packages.Nikola</li>
    <li>python310Packages.arviz</li>
    <li>python310Packages.augmax</li>
    <li>python310Packages.batchspawner</li>
    <li>python310Packages.dalle-mini</li>
    <li>python310Packages.dockerspawner</li>
    <li>python310Packages.elegy</li>
    <li>python310Packages.jupyterhub</li>
    <li>python310Packages.jupyterhub-ldapauthenticator</li>
    <li>python310Packages.jupyterhub-systemdspawner</li>
    <li>python310Packages.jupyterhub-tmpauthenticator</li>
    <li>python310Packages.n3fit</li>
    <li>python310Packages.napari</li>
    <li>python310Packages.napari-npe2</li>
    <li>python310Packages.oauthenticator</li>
    <li>python310Packages.pymc</li>
    <li>python310Packages.treex</li>
    <li>python310Packages.validphys2</li>
    <li>python311Packages.Nikola</li>
    <li>python311Packages.augmax</li>
    <li>python311Packages.batchspawner</li>
    <li>python311Packages.dockerspawner</li>
    <li>python311Packages.jupyterhub</li>
    <li>python311Packages.jupyterhub-ldapauthenticator</li>
    <li>python311Packages.jupyterhub-systemdspawner</li>
    <li>python311Packages.jupyterhub-tmpauthenticator</li>
    <li>python311Packages.oauthenticator</li>
    <li>python311Packages.validphys2</li>
    <li>stack2nix</li>
    <li>stutter</li>
    <li>stylish-cabal</li>
    <li>tamarin-prover</li>
    <li>taskell</li>
    <li>tlaps</li>
    <li>tweet-hs</li>
    <li>uuagc</li>
    <li>uusi</li>
    <li>validphys2</li>
  </ul>
</details>
<details>
  <summary>19 packages failed to build:</summary>
  <ul>
    <li>fishPlugins.fishtape_3</li>
    <li>fishPlugins.fzf-fish</li>
    <li>gitit</li>
    <li>hledger-check-fancyassertions</li>
    <li>jacinda</li>
    <li>klipper-genconf</li>
    <li>nix-serve-ng</li>
    <li>python310Packages.ansible-kernel</li>
    <li>python310Packages.jupyter-server-mathjax</li>
    <li>python310Packages.jupyterlab-git</li>
    <li>python310Packages.nbdime</li>
    <li>python311Packages.ansible-kernel</li>
    <li>python311Packages.httpie</li>
    <li>python311Packages.httpie-ntlm</li>
    <li>python311Packages.jupyter-server-mathjax</li>
    <li>python311Packages.jupyterlab-git</li>
    <li>python311Packages.nbdime</li>
    <li>spacecookie</li>
    <li>trenchbroom</li>
  </ul>
</details>
<details>
  <summary>179 packages built:</summary>
  <ul>
    <li>actionlint</li>
    <li>agda</li>
    <li>alex</li>
    <li>apksigcopier</li>
    <li>arion</li>
    <li>bashblog</li>
    <li>bench</li>
    <li>bgnet</li>
    <li>blucontrol</li>
    <li>bup</li>
    <li>cabal-install</li>
    <li>cabal2nix</li>
    <li>cabal2nix-unwrapped</li>
    <li>cachix</li>
    <li>carp</li>
    <li>client-ip-echo</li>
    <li>coursera-dl</li>
    <li>darcs</li>
    <li>darcs-to-git</li>
    <li>dconf2nix</li>
    <li>dhall</li>
    <li>dhall-bash</li>
    <li>dhall-docs</li>
    <li>dhall-json</li>
    <li>dhall-lsp-server</li>
    <li>dhall-nix</li>
    <li>dhall-nixpkgs</li>
    <li>dhallPackages.Prelude</li>
    <li>dhallPackages.dhall-cloudformation</li>
    <li>dhallPackages.dhall-grafana</li>
    <li>dhallPackages.dhall-kubernetes</li>
    <li>dogdns</li>
    <li>echidna</li>
    <li>eget</li>
    <li>elm2nix</li>
    <li>emacsPackages.agda2-mode</li>
    <li>emacsPackages.hindent</li>
    <li>emanote</li>
    <li>exa</li>
    <li>fend</li>
    <li>fffuu</li>
    <li>fishPlugins.bass</li>
    <li>fishPlugins.done</li>
    <li>fishPlugins.fishtape</li>
    <li>futhark</li>
    <li>gg-scm</li>
    <li>ghc</li>
    <li>ghcid</li>
    <li>git-annex</li>
    <li>git-brunch</li>
    <li>git-ftp</li>
    <li>git-open</li>
    <li>glirc</li>
    <li>gocryptfs</li>
    <li>hadolint</li>
    <li>happy</li>
    <li>haskell-ci</li>
    <li>haskell-language-server</li>
    <li>hci</li>
    <li>hercules-ci-agent</li>
    <li>hinit</li>
    <li>hledger</li>
    <li>hledger-interest</li>
    <li>hledger-ui</li>
    <li>hledger-utils (python310Packages.hledger-utils)</li>
    <li>hledger-web</li>
    <li>hlint</li>
    <li>hpack</li>
    <li>hscolour</li>
    <li>http-prompt</li>
    <li>httpie (python310Packages.httpie)</li>
    <li>idris</li>
    <li>ihaskell</li>
    <li>isabelle</li>
    <li>isabelle-components.isabelle-linter</li>
    <li>jl</li>
    <li>jo</li>
    <li>json2yaml</li>
    <li>jupyter</li>
    <li>koka</li>
    <li>krank</li>
    <li>lambdabot</li>
    <li>ledger-autosync</li>
    <li>lhs2tex</li>
    <li>lsd</li>
    <li>matterhorn</li>
    <li>memo</li>
    <li>mp3fs</li>
    <li>mpdcron</li>
    <li>mueval</li>
    <li>naproche</li>
    <li>niv</li>
    <li>nix-deploy</li>
    <li>nix-diff</li>
    <li>nix-info-tested</li>
    <li>nix-output-monitor</li>
    <li>nix-script</li>
    <li>nix-tree</li>
    <li>nixfmt</li>
    <li>nominatim</li>
    <li>nota</li>
    <li>nvfetcher</li>
    <li>ormolu</li>
    <li>osmium-tool</li>
    <li>ov</li>
    <li>pandoc</li>
    <li>passExtensions.pass-checkup</li>
    <li>pinboard-notes-backup</li>
    <li>place-cursor-at</li>
    <li>pretty-simple</li>
    <li>purenix</li>
    <li>python310Packages.einops</li>
    <li>python310Packages.git-annex-adapter</li>
    <li>python310Packages.httpie-ntlm</li>
    <li>python310Packages.jupyter</li>
    <li>python310Packages.jupyter-lsp</li>
    <li>python310Packages.jupyter-server</li>
    <li>python310Packages.jupyterlab</li>
    <li>python310Packages.jupyterlab-lsp</li>
    <li>python310Packages.jupyterlab_launcher</li>
    <li>python310Packages.jupyterlab_server</li>
    <li>python310Packages.jupytext</li>
    <li>python310Packages.mkdocs-jupyter</li>
    <li>python310Packages.nbclassic</li>
    <li>python310Packages.nix-kernel</li>
    <li>python310Packages.notebook</li>
    <li>python310Packages.notebook-shim</li>
    <li>python310Packages.notedown</li>
    <li>python310Packages.pelican</li>
    <li>python310Packages.phik</li>
    <li>python310Packages.pymoo</li>
    <li>python310Packages.pypandoc</li>
    <li>python310Packages.pytm</li>
    <li>python310Packages.pytomlpp</li>
    <li>python310Packages.reportengine</li>
    <li>python310Packages.vector</li>
    <li>python310Packages.vega</li>
    <li>python310Packages.xarray-einstats</li>
    <li>python311Packages.einops</li>
    <li>python311Packages.git-annex-adapter</li>
    <li>python311Packages.hledger-utils</li>
    <li>python311Packages.jupyter</li>
    <li>python311Packages.jupyter-lsp</li>
    <li>python311Packages.jupyter-server</li>
    <li>python311Packages.jupyterlab</li>
    <li>python311Packages.jupyterlab-lsp</li>
    <li>python311Packages.jupyterlab_launcher</li>
    <li>python311Packages.jupyterlab_server</li>
    <li>python311Packages.jupytext</li>
    <li>python311Packages.mkdocs-jupyter</li>
    <li>python311Packages.nbclassic</li>
    <li>python311Packages.nix-kernel</li>
    <li>python311Packages.notebook</li>
    <li>python311Packages.notebook-shim</li>
    <li>python311Packages.notedown</li>
    <li>python311Packages.pelican</li>
    <li>python311Packages.pypandoc</li>
    <li>python311Packages.pytm</li>
    <li>python311Packages.pytomlpp</li>
    <li>python311Packages.reportengine</li>
    <li>python311Packages.vega</li>
    <li>ripgrep-all</li>
    <li>scriv</li>
    <li>shake</li>
    <li>shellcheck</li>
    <li>splot</li>
    <li>stack</li>
    <li>stylish-haskell</li>
    <li>tldr-hs</li>
    <li>update-nix-fetchgit</li>
    <li>vaultenv</li>
    <li>vimPlugins.cmp-pandoc-nvim</li>
    <li>vimPlugins.vim-stylish-haskell</li>
    <li>vscode-extensions.timonwong.shellcheck</li>
    <li>wayback-machine-archiver</li>
    <li>wstunnel</li>
    <li>xmonad-with-packages</li>
    <li>xmonadctl</li>
    <li>zsh-git-prompt</li>
  </ul>
</details>

File: pkgs/development/compilers/ghc/cabal-paths-9.2.x.patch (hunk 1 above)
Comment (matthewess) [reply pkgs/development/compilers/ghc/cabal-paths-9.2.x.patch]: Resolved this with a variation on the top patch, which gets rid of splitPath entirely.

**Comment by matthewess** [comment]
@srid @domenkozar if either of you has an x86_64-darwin machine to test this PR, I think that would help make it a bit more convincing for merging. `nixpkgs-review`, but primarily running `ormolu` or `ghcid`.

for reference, when I try to run `ghcid` on my `aarch64-darwin` machine, I get this failure:
```bash
❯ nix run .#haskell.packages.ghc924.ghcid.bin
Loading ghci -fno-code -fno-break-on-exception -fno-break-on-error -v1 -ferror-spans -j ...
GHCi, version 9.2.4: https://www.haskell.org/ghc/  :? for help

No files loaded, GHCi is not working properly.
Command: ghci -fno-code -fno-break-on-exception -fno-break-on-error -v1 -ferror-spans -j
```
can't tell if this is due to the patch, or some unrelated issue.

**Comment by matthewess** [comment]
similarly strange error for ormolu: 

```bash
❯ nix run .#haskell.packages.ghc924.ormolu
error: unable to execute '/nix/store/1y4bg6lk0qqlrpqhnxw1j83bgw5c2rds-ormolu-0.5.0.1/bin/ormolu': No such file or directory

❯ ls /nix/store/1y4bg6lk0qqlrpqhnxw1j83bgw5c2rds-ormolu-0.5.0.1/
lib  nix-support
```

**Comment by sternenseemann** [comment]
I think we can actually make it even simple, I'll try to come up with an alternative patch and a way to test this that doesn't require rebuilding GHC everytime which will probably come in handy.

**Comment by sternenseemann** [comment]
@matthewess What do you think of [this](https://github.com/sternenseemann/cabal/compare/982646d67b95b32813b89ab5d2d2f4d4dc03fb2b..sterni-Cabal-3.6-for-aarch64-darwin)? I tried moving all code generation logic into the Template, so that only the decision whether stuff is emitted or not (and the generation of the warning text) is left in the `PathsModule` module. This way the amount of weird code in there is limited and things are hopefully a bit clearer or easier to understand thanks to the separation.

**Comment by matthewess** [comment]
@sternenseemann that looks very clean! i can try and work those changes into this PR this weekend

**Comment by sternenseemann** [comment]
@matthewess If you want, I can take that off your hands; then I'd just need your help with reviewing and testing (since I lack the hardware).

**Comment by sternenseemann** [comment]
@ofborg build haskell.compiler.ghc927 haskell.compiler.ghc926

For anyone reviewing, viewing the [patch this way](https://github.com/sternenseemann/cabal/compare/982646d67b95b32813b89ab5d2d2f4d4dc03fb2b..e925c8cbe3011a7d83d1184e4cd89afd17a368b6) may be easier.

**Comment by sternenseemann** [comment]
@ofborg build haskell.compiler.ghc927 haskell.compiler.ghc926

**Comment by sternenseemann** [comment]
Needs to be tested still. I think it's okay, but would be preferable if someone else could also review the actual code.

**Comment by matthewess** [comment]
Playing around with this today. Are there any good test candidates besides making sure that e.g. `ormolu` and `ghcid` will run?

**Comment by sternenseemann** [comment]
@matthewess plus `niv`, `nfc` and `citeproc`

File: pkgs/development/compilers/ghc/Cabal-3.6-paths-fix-cycle-aarch64-darwin.patch [hunk 2]
```diff
@@ -0,0 +1,989 @@
...
+        , ("SysconfDir", flat_sysconfdir)
+        ]
+ 
++    shouldEmit :: String -> Bool
+     shouldEmit name =
+       case lookup name dirs of
+         Just b -> b
+         Nothing -> error "panic! BUG in Cabal Paths_ patch for aarch64-darwin, report this at https://github.com/nixos/nixpkgs/issues"
+ 
++    -- This is a comma separated list of all functions that have been emitted.
```
Comment (maralorn) [inline pkgs/development/compilers/ghc/Cabal-3.6-paths-fix-cycle-aarch64-darwin.patch:969]: This comment should say "omitted", right?

**Review by maralorn — APPROVED** [review]
I cannot claim that I have understood this change in full depth. Especially since I hadn’t looked at the Cabal code base before.

Still, this change is well documented, and the code looks like it does what the comment says. Also, to prove that I really read it, I found a subtle typo.

**Comment by matthewess** [comment]
The following commands worked for me on `aarch64-darwin`:

```bash
❯ nix run .#haskell.packages.ghc924.ormolu.bin
❯ nix run .#haskell.packages.ghc924.ghcid.bin -- --help
❯ nix run .#haskell.packages.ghc924.niv.bin
❯ nix run .#haskell.packages.ghc924.citeproc.bin -- --help
```

attempting to run `haskell.packages.ghc924.nfc[.bin]` threw an error, and when I inspected the nix store folder, I saw the same no-`bin` contents as here: https://github.com/NixOS/nixpkgs/pull/216857#issuecomment-1436120281

**Comment by sternenseemann** [comment]
@matthewess The thing for ormolu is to be expected as you were not actually looking at the `bin` output when listing the store path. Not sure why `nix run` misbehaves here, no experience with the flakes CLI's behavior.

File: pkgs/development/compilers/ghc/Cabal-3.6-paths-fix-cycle-aarch64-darwin.patch (hunk 2 above)
Comment (sternenseemann) [reply pkgs/development/compilers/ghc/Cabal-3.6-paths-fix-cycle-aarch64-darwin.patch]: Yes, fixed!

**Comment by matthewess** [comment]
> @matthewess The thing for ormolu is to be expected as you were not actually looking at the `bin` output when listing the store path. Not sure why `nix run` misbehaves here, no experience with the flakes CLI's behavior.

Good catch, you can see I figured that out for ormolu but I messed up with `nfc`. I think it must only have one oddly named binary?

```bash
❯ nix run .#haskell.packages.ghc924.nfc.bin
error: unable to execute '/nix/store/pd62xcxz72g335axy6872002fihq9qp1-nfc-0.1.1-bin/bin/nfc': No such file or directory

❯ ls /nix/store/pd62xcxz72g335axy6872002fihq9qp1-nfc-0.1.1-bin/bin/
print-mifare-uid-forever
```

**Comment by sternenseemann** [comment]
@matthewess So GHC is still working on aarch64-darwin? Should we merge?

**Comment by matthewess** [comment]
> @matthewess So GHC is still working on aarch64-darwin? Should we merge?

Yes, GHC is working! My somewhat informed opinion on merging is also yes. I'm no Haskell maintainer, but this seems to fix the symptom that prompted the original PR well, and your changes set us up better to adjust it in the future if something ends up being amiss.

**Pull request merged**